faucet/internal/failure/failure.go

```diff
@@ -0,0 +1,89 @@
+package failure
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/faucet/internal/platform/lotus"
+)
+
+type Detector struct {
+	m                         sync.Mutex
+	log                       *logging.ZapEventLogger
+	lastBlockHeight           uint64
+	lotus                     lotus.API
+	lastBlockHeightUpdateTime time.Time
+	threshold                 time.Duration
+	checkInterval             time.Duration
+	stopChan                  chan bool
+	ticker                    *time.Ticker
+}
+
+// NewDetector creates a new failure detector that checks height value each checkInterval
+// and triggers a failure if there is no block height update in threshold.
+func NewDetector(log *logging.ZapEventLogger, api lotus.API, checkInterval, threshold time.Duration) *Detector {
+	d := Detector{
+		checkInterval:             checkInterval,
+		lotus:                     api,
+		log:                       log,
+		threshold:                 threshold,
+		stopChan:                  make(chan bool),
+		lastBlockHeightUpdateTime: time.Now(),
+		ticker:                    time.NewTicker(checkInterval),
+	}
+
+	go d.run()
+
+	return &d
+}
+
+func (d *Detector) run() {
+	for {
+		select {
+		case <-d.stopChan:
+			close(d.stopChan)
+			d.log.Infow("shutdown", "status", "detector stopped")
+			return
+		case <-d.ticker.C:
+			status, err := d.lotus.NodeStatus(context.Background(), true)
+			if err != nil {
+				d.log.Errorw("error", "detector", "unable to get block", err)
+			} else {
+				height := status.SyncStatus.Epoch
+				d.m.Lock()
+				if d.lastBlockHeight != height {
+					d.lastBlockHeight = height
+					d.lastBlockHeightUpdateTime = time.Now()
+				}
+				d.m.Unlock()
+			}
+		}
+	}
+
+}
+
+func (d *Detector) Stop() {
+	d.ticker.Stop()
+	d.stopChan <- true
+}
+
+func (d *Detector) GetLastBlockHeight() uint64 {
+	d.m.Lock()
+	defer d.m.Unlock()
+	return d.lastBlockHeight
+}
+
+func (d *Detector) IsFailed() error {
```

Comment:
Cosmetic nit: I'd either rename this to something like `CheckProgress`, or return a `bool`. That would be more intuitive for me at least.

Reply:
I renamed the function to `CheckProgress`, but I would prefer to return an error because I return information related to the last block and last block time in the error.

Reply:
Yes, for `CheckProgress`, returning an `error` is completely intuitive, this is exactly how I meant it.